daisy/workflow/step_wait_for_instances_signal.go

```diff
@@ -14,20 +14,125 @@
+	FailureMatch string
+}
+
+// InstanceSignal waits for a signal from an instance.
+type InstanceSignal struct {
+	// Instance name to wait for.
+	Name string
+	// Interval to check for signal (default is 5s).
+	Interval time.Duration
+	// Wait for the instance to stop.
+	Stopped bool
+	// Wait for a string match in the serial output.
+	SerialOutput *SerialOutput
+}
+
+func waitForSerialOutput(w *Workflow, name string, port int64, success, failure string, interval time.Duration) error {
+	w.logger.Printf("WaitForInstancesSignal: watching serial port %d, SuccessMatch: %q, FailureMatch: %q.", port, success, failure)
+	var start int64
+	for {
+		tick := time.Tick(interval)
+		select {
+		case <-w.Ctx.Done():
+		case <-tick:
+			resp, err := w.ComputeClient.GetSerialPortOutput(w.Project, w.Zone, name, port, start)
+			if err != nil {
+				stopped, sErr := w.ComputeClient.InstanceStopped(w.Project, w.Zone, name)
+				if stopped && sErr == nil {
+					w.logger.Printf("WaitForInstancesSignal: instance %q stopped, not waiting for serial output.", name)
+					return nil
+				}
+				return fmt.Errorf("WaitForInstancesSignal: instance %q: error getting serial port: %v", name, err)
+			}
+			start = resp.Next
+			if strings.Contains(resp.Contents, success) {
+				w.logger.Printf("WaitForInstancesSignal: SuccessMatch instance %q: %q in %q", name, failure, resp.Contents)
+				return nil
+			}
+			if failure != "" && strings.Contains(resp.Contents, failure) {
+				return fmt.Errorf("WaitForInstancesSignal: FailureMatch instance %q: %q in %q", name, failure, resp.Contents)
+			}
+		}
+	}
+}
 
 func (s *WaitForInstancesSignal) run(w *Workflow) error {
-	return nil
+	var wg sync.WaitGroup
+	e := make(chan error)
+
+	for _, is := range *s {
+		if is.Interval == 0 {
+			is.Interval = 5 * time.Second
+		}
+		wg.Add(1)
+		go func(is InstanceSignal) {
+			defer wg.Done()
+			i, ok := w.instanceRefs.get(is.Name)
+			if !ok {
+				e <- fmt.Errorf("unresolved instance %q", is.Name)
+				return
+			}
+			if is.Stopped {
+				w.logger.Printf("WaitForInstancesSignal: waiting for instance %q to stop.", i.real)
+				if err := w.ComputeClient.WaitForInstanceStopped(w.Project, w.Zone, i.real, is.Interval); err != nil {
+					e <- err
+					return
+				}
+				w.logger.Printf("WaitForInstancesSignal: instance %q stopped.", i.real)
+			}
+			if is.SerialOutput != nil {
+				if err := waitForSerialOutput(w, i.real, is.SerialOutput.Port, is.SerialOutput.SuccessMatch, is.SerialOutput.FailureMatch, is.Interval); err != nil {
+					e <- err
+					return
+				}
+			}
+		}(is)
+	}
+
+	go func() {
+		wg.Wait()
+		e <- nil
+	}()
+
+	select {
+	case err := <-e:
+		return err
+	case <-w.Cancel:
+		return nil
+	}
 }
 
 func (s *WaitForInstancesSignal) validate(w *Workflow) error {
 	// Instance checking.
 	for _, i := range *s {
-		if !instanceValid(w, i) {
-			return fmt.Errorf("cannot wait for instance signal. Instance not found: %s", i)
+		if !instanceValid(w, i.Name) {
```

Comment:
If we can get rid of instanceValid, we don't have to change it in this PR; file a bug against me.